src/query/api/v1/handler/prometheus/remote/write.go

```diff
@@ -509,3 +510,17 @@ func (i *promTSIter) Reset() error {
 func (i *promTSIter) Error() error {
 	return nil
 }
+
+func (i *promTSIter) SetCurrentMetadata(metadata ts.Metadata) {
+	if len(i.metadatas) == 0 {
+		i.metadatas = make([]ts.Metadata, len(i.tags))
+	}
+	i.metadatas[i.idx] = metadata
+}
+
+func (i *promTSIter) CurrentMetadata() ts.Metadata {
+	if len(i.metadatas) == 0 {
+		return ts.Metadata{}
+	}
+	return i.metadatas[i.idx]
+}
```

Comment:
Rather than making this slice, maybe have a `currentMetadata` and `metaIndex` field on `promTSIter` and these methods look like:

```
func (i *promTSIter) SetCurrentMetadata(metadata ts.Metadata) {
 if i.idx < 0 {
  return
 }
 i.currentMetadata := metadata
 i.metaIndex := i.idx
}
 
func (i *promTSIter) CurrentMetadata() ts.Metadata {
 if i.idx == i.metaIndex {
  return i.currentMetadata
 }

 return ts.Metadata{}
}
```